pkg/servercounter/lease_counter.go

```diff
@@ -0,0 +1,73 @@
+package servercounter
+
+import (
+	"fmt"
+	"time"
+
+	coordinationv1api "k8s.io/api/coordination/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	coordinationv1listers "k8s.io/client-go/listers/coordination/v1"
+	"k8s.io/klog/v2"
+)
+
+type ServerLeaseCounter struct {
+	lister        coordinationv1listers.LeaseLister
+	selector      labels.Selector
+	fallbackCount int
+}
+
+func NewServerLeaseCounter(lister coordinationv1listers.LeaseLister, labelSelector string, fallbackCount int) (*ServerLeaseCounter, error) {
+	selector, err := labels.Parse(labelSelector)
+	if err != nil {
+		return nil, fmt.Errorf("could not parse label selector %v: %w", labelSelector, err)
+	}
+	return &ServerLeaseCounter{
+		lister:        lister,
+		selector:      selector,
+		fallbackCount: fallbackCount,
+	}, nil
+}
+
+func (lc *ServerLeaseCounter) CountServers() int {
+	leases, err := lc.lister.List(lc.selector)
+	if err != nil {
+		klog.Errorf("could not list leases to update server count, using fallback count (%v): %v", lc.fallbackCount, err)
+		return lc.fallbackCount
```

Comment:
We need to consider the implications of the List failure.  In general we can't assume kube-apiserver will always send a successful response.  If we get a single List failure, should we immediately return the fallback value?  What behavior would that cause in the agent sync loop?

Reply:
Oh I didn't read the block before writing this comment

```
        if count != lc.fallbackCount {
		klog.Infof("found %v valid leases, updating fallback count", count)
		lc.fallbackCount = count
	}
```

 I think the behavior is fine. Could you add a comment here briefly explaining the agent behavior in the event of List failures however?

Reply:
Added to the function doc comment.